call-cabal-to-nix.nix

```diff
@@ -0,0 +1,78 @@
+{ stdenv, lib, buildPackages, mkPkgSet, nix-tools }:
+
+let
+
+  haskellSrc2nix = { name, src ? null, extraCabal2nixOptions ? "" }:
+    buildPackages.stdenv.mkDerivation {
+      name = "cabal-to-nix-${name}";
+      nativeBuildInputs = [ nix-tools ];
+      preferLocalBuild = true;
+      allowSubstitutes = false;
+      phases = ["installPhase"];
+      LANG = "en_US.UTF-8";
+      LOCALE_ARCHIVE = lib.optionalString (stdenv.buildPlatform.libc == "glibc") "${buildPackages.glibcLocales}/lib/locale/locale-archive";
+      installPhase = ''
+        export HOME="$TMP"
+        mkdir -p "$out"
+        cabal-to-nix "${src}/${name}.cabal" ${extraCabal2nixOptions} > "$out/default.nix"
+      '';
+    };
+
+  haskellSrc2plan = { name, src, ghc }:
```

Comment:
> This is inherently impure and will yield different results for the same derivation if run at different times.

Yeah, I know. But if it's possible to somehow pin the Hackage index, then it would be deterministic.

What I would like is if the `hackage.nix` regeneration script also produces a fixed-output `fetchurl` derivation for `01-index.tar.gz`. If `fetchurl` could specify a HTTP Range header or truncate its result then this would be easy.

So does anyone have any ideas how to achieve this?

Reply:
@rvl In theory, deleting all entries from the tarfile that are newer than an exact timestamp from the past should always yield the same result, since the tarball is only ever changed by adding new entries with new dates.

Reply:
Might want to ask someone from the Hackage crew how reliable that would be though.

Reply:
If we include this information #62 in `hackage.nix` then we can use the index-state feature: `cabal new-update hackage.haskell.org,2016-09-24T17:47:48Z`. But that still relies on network access.

To build in a sandbox we need a fixed-output derivation for the [package index](http://hackage.haskell.org/api#core). I'm not sure how to get that -- it's more of a Nix problem.

Reply:
you *can tell* cabal to update to a specific index state. That should give you reproducibility.

Reply:
Does that actually restrict the entries of the tarball? Or does it just instruct cabal-install to ignore newer entries? We need a restricted tarball file. Basically we'd have one fixed-output derivation that fetches the tarball and makes sure it's only got entries up to a certain time. Then cabal new-configure can be done in a regular derivation to get the plan.

Reply:
*(Heck, if we rewrote plan-to-nix **in nix**, we could `builtins.fromJSON (builtins.readFile (pkgs.runCommand "run-cabal-new-configure" ...))` the resulting plan.json file to produce a plan at eval time :P)*

Reply:
My understanding of the index state is that it should give you a reproducible index.

@hvr, care to enlighten us?

Reply:
So, @hvr told me the following:

if we fix the environment deterministically and fix the index-state timestamp the output of cabal will be deterministic. The index-state (via `--index-state=`) gives you a unique hackage snapshot, that is fixed in time.

Reply:
@angerman As far as I can tell, the only thing that's reproducible is cabal's behavior with respect to the index. After experimenting with trying to downgrade the tarball manually, it doesn't look like the entries are actually removed. Thus the tarball itself is not reproducible with cabal new-update; you will always get a tarball populated with all packages currently on Hackage, and cabal-install just knows how to ignore the ones newer than the index-state

Reply:
That's fine for cabal-install, but it's not good enough for Nix. We need to know the hash of the tarball in order to use the network and download it so that Nix can verify it was the exact tarball we wanted. i.e. we need to delete entries from the tarball after downloading it.

Reply:
So we'll need some tool to filter `tar` files. Great :-/ 
A pure nix plan-to-nix will not be able to deal with full `plan.json` files I'm afraid,
as you'd need a pure cabal-to-nix one as well.

Reply:
@angerman Luckily it's extremely easy to write such a program. [The `Entries` type](http://hackage.haskell.org/package/tar-0.5.1.0/docs/Codec-Archive-Tar.html#t:Entries) is pretty easy to manipulate.

Also, we could even use this to eliminate the need for the hackage.nix repo. Since we'll have a Nix expression for the tarball anyway, we can have a Nix expression for running hackage-to-nix on it too.

Reply:
That sounds easy enough. Dropping hackage.nix in favour of generating it on demand sounds also tempting. But we might need to speed that process up in that case (or provide the necessary cache on some instance; last I checked it did take quite some time).